File: ports/typescript/src/generator.ts
/**
 * Generate the HTML dashboard from parsed data.
 */

import * as fs from "fs";
import * as path from "path";

import type { ParsedData } from "./parser";
import type { RecommendationResult } from "./analyzer";

// ============================================================
// Template loading
// ============================================================

function getTemplate(): string {
  // Look for template.html relative to this file (works both in src/ and dist/)
  const candidates = [
    path.join(__dirname, "..", "template.html"),
    path.join(__dirname, "..", "..", "template.html"),
    path.join(process.cwd(), "template.html"),
  ];
  for (const p of candidates) {
    if (fs.existsSync(p)) {
      return fs.readFileSync(p, "utf-8");
    }
  }
  throw new Error(
    `Template not found. Searched: ${candidates.join(", ")}`
  );
}

// ============================================================
// HTML generation
// ============================================================

export function generateHtml(
  data: ParsedData,
  recommendations: RecommendationResult
): string {
  const template = getTemplate();

  const payload: Record<string, unknown> = {
    dashboard: data.dashboard,
    drilldown: data.drilldown,
    analysis: data.analysis,
    recommendations,
    work_days: data.work_days || [],
    models: data.models || [],
    subagents: data.subagents || {},
    branches: data.branches || [],
    context_efficiency: data.context_efficiency || {},
    versions: data.versions || [],
    skills: data.skills || [],
    slash_commands: data.slash_commands || [],
    config: data.config || {},
  };

  const payloadJson = JSON.stringify(payload);
  const html = template.replace("__DATA_PLACEHOLDER__", payloadJson);
  return html;
}

// ============================================================
// Report writing
// ============================================================

export function writeReport(html: string, outputPath?: string | null): string {
  let finalPath: string;

  if (!outputPath) {
    const outputDir = path.join(process.cwd(), "output");
    fs.mkdirSync(outputDir, { recursive: true });

    const now = new Date();
    const timestamp = [
      now.getFullYear(),
      String(now.getMonth() + 1).padStart(2, "0"),
      String(now.getDate()).padStart(2, "0"),
      "-",
      String(now.getHours()).padStart(2, "0"),
      String(now.getMinutes()).padStart(2, "0"),
      String(now.getSeconds()).padStart(2, "0"),
    ].join("");
    finalPath = path.join(outputDir, `claude-analytics-${timestamp}.html`);

    ensureGitignore(outputDir);
  } else {
    finalPath = outputPath;
  }

  const dir = path.dirname(finalPath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(finalPath, html, "utf-8");

  return finalPath;
}

// ============================================================
// Last-run tracking
// ============================================================

export function readLastRun(outputDir?: string): string | null {
  const dir = outputDir || path.join(process.cwd(), "output");
  const marker = path.join(dir, ".last-run");
  if (fs.existsSync(marker)) {
    const content = fs.readFileSync(marker, "utf-8").trim();
    return content.length >= 10 ? content.substring(0, 10) : null;
  }
  return null;
}

export function saveLastRun(outputDir?: string): void {
  const dir = outputDir || path.join(process.cwd(), "output");
  fs.mkdirSync(dir, { recursive: true });
  const marker = path.join(dir, ".last-run");
  const now = new Date();
  const ts = [
    now.getFullYear(),
    "-",
    String(now.getMonth() + 1).padStart(2, "0"),
    "-",
    String(now.getDate()).padStart(2, "0"),
    " ",
    String(now.getHours()).padStart(2, "0"),
    ":",
    String(now.getMinutes()).padStart(2, "0"),
    ":",
    String(now.getSeconds()).padStart(2, "0"),
  ].join("");
  fs.writeFileSync(marker, ts + "\n", "utf-8");
}

// ============================================================
// Gitignore management
// ============================================================

function ensureGitignore(outputDir: string): void {
  const innerGitignore = path.join(outputDir, ".gitignore");
  if (!fs.existsSync(innerGitignore)) {
    fs.writeFileSync(
      innerGitignore,
      "# Claude Analytics reports contain sensitive session data.\n" +
        "# Do NOT commit these files to version control.\n" +
        "*\n" +
        "!.gitignore\n",
      "utf-8"
    );
  }

  const rootGitignore = path.join(path.dirname(outputDir), ".gitignore");
  if (fs.existsSync(rootGitignore)) {
    const content = fs.readFileSync(rootGitignore, "utf-8");
    if (!content.includes("output/")) {
      fs.appendFileSync(
        rootGitignore,
        "\n# Claude Analytics reports (contain sensitive data)\noutput/\n",
        "utf-8"
      );
    }
  } else {
    fs.writeFileSync(
      rootGitignore,
      "# Claude Analytics reports (contain sensitive data)\noutput/\n",
      "utf-8"
    );
  }
}

File: ports/python/claude_analytics/generator.py
"""Generate the HTML dashboard from parsed data."""

import json
import os
from datetime import datetime
from pathlib import Path


def get_template():
    """Load the HTML template from the package directory."""
    template_path = Path(__file__).parent / "template.html"
    if not template_path.exists():
        raise FileNotFoundError(f"Template not found at {template_path}")
    return template_path.read_text(encoding="utf-8")


def generate_html(data, recommendations):
    """Generate the final HTML dashboard.

    Args:
        data: Parsed session data from parser.parse_all_sessions()
        recommendations: Dict from analyzer.generate_recommendations()

    Returns:
        str: Complete HTML content
    """
    template = get_template()

    # Build the payload: everything the template needs
    payload = {
        "dashboard": data["dashboard"],
        "drilldown": data["drilldown"],
        "analysis": data["analysis"],
        "recommendations": recommendations,
        "work_days": data.get("work_days", []),
        # New data
        "models": data.get("models", []),
        "subagents": data.get("subagents", {}),
        "branches": data.get("branches", []),
        "context_efficiency": data.get("context_efficiency", {}),
        "versions": data.get("versions", []),
        "skills": data.get("skills", []),
        "slash_commands": data.get("slash_commands", []),
        "config": data.get("config", {}),
    }

    # Inject data into template
    payload_json = json.dumps(payload, separators=(",", ":"))
    html = template.replace("__DATA_PLACEHOLDER__", payload_json)

    return html


def write_report(html, output_path=None):
    """Write the HTML report to disk.

    Reports are saved to an 'output/' directory with timestamped filenames
    so previous runs are preserved for tracking improvement over time.
    The output/ folder is automatically .gitignored to prevent leaking
    sensitive session data.

    Args:
        html: HTML content string
        output_path: Where to write. Defaults to ./output/claude-analytics-YYYYMMDD-HHMMSS.html

    Returns:
        Path: The output file path
    """
    if output_path is None:
        output_dir = Path.cwd() / "output"
        output_dir.mkdir(parents=True, exist_ok=True)
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        output_path = output_dir / f"claude-analytics-{timestamp}.html"

        # Ensure output/ is gitignored
        _ensure_gitignore(output_dir)
    else:
        output_path = Path(output_path)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(html, encoding="utf-8")

    return output_path


def read_last_run(output_dir=None):
    """Read the last run timestamp from the .last-run marker file.

    Returns:
        str or None: ISO date string (YYYY-MM-DD) of the last run, or None.
    """
    if output_dir is None:
        output_dir = Path.cwd() / "output"
    marker = output_dir / ".last-run"
    if marker.exists():
        content = marker.read_text(encoding="utf-8").strip()
        # Return just the date portion
        return content[:10] if len(content) >= 10 else None
    return None


def save_last_run(output_dir=None):
    """Save the current timestamp as the last run marker.

    Args:
        output_dir: The output directory. Defaults to ./output/
    """
    if output_dir is None:
        output_dir = Path.cwd() / "output"
    output_dir.mkdir(parents=True, exist_ok=True)
    marker = output_dir / ".last-run"
    marker.write_text(
        datetime.now().strftime("%Y-%m-%d %H:%M:%S") + "\n",
        encoding="utf-8",
    )


def _ensure_gitignore(output_dir):
    """Create or update .gitignore in the output directory.

    Also adds 'output/' to the project root .gitignore if it exists.
    """
    # .gitignore inside output/ — catch-all
    inner_gitignore = output_dir / ".gitignore"
    if not inner_gitignore.exists():
        inner_gitignore.write_text(
            "# Claude Analytics reports contain sensitive session data.\n"
            "# Do NOT commit these files to version control.\n"
            "*\n"
            "!.gitignore\n",
            encoding="utf-8",
        )

    # Also add 'output/' to the project root .gitignore
    root_gitignore = output_dir.parent / ".gitignore"
    if root_gitignore.exists():
        content = root_gitignore.read_text(encoding="utf-8")
        if "output/" not in content:
            with open(root_gitignore, "a", encoding="utf-8") as f:
                f.write("\n# Claude Analytics reports (contain sensitive data)\noutput/\n")
    else:
        root_gitignore.write_text(
            "# Claude Analytics reports (contain sensitive data)\noutput/\n",
            encoding="utf-8",
        )
